lib/Interface.h

```
@@ -0,0 +1,148 @@
+        return false;
+    }
+
+    // Copy content from source to destination
+    destinationFile << sourceFile.rdbuf();
+
+    if (destinationFile.fail()) {
+        std::cerr << "Error copying file." << std::endl;
+        return false;
+    }
+    sourceFile.close();
+    destinationFile.close();
+
+    return true;
+}
+// #############################################################################
+//                           DRILModule
+// #############################################################################
+
+template <typename E, size_t NumSymbols>
+class DRILModule
+{
+public:
+  static void Loadlibrary() { GetInstance().Load(1); }
+  static void Reloadlibrary() { GetInstance().Reload(); }
+protected:
+  static E& GetInstance()
+  {
+    static E instance;
+    return instance;
+  }
+
+  // Should return the path to the library on disk
+  virtual const char* GetPath() const = 0;
+
+  // Should return a reference to an array of C-strings of size NumSymbols
+  // Used when loading or reloading the library to lookup the address of
+  // all exported symbols
+  virtual std::array<const char*, NumSymbols>& GetSymbolNames() const = 0;
+
+  template <typename Ret, typename... Args>
+  Ret Execute(const char* name, Args... args)
+  {
+    // Lookup the function address
+    auto symbol = m_symbols.find(name);
+    if (symbol != m_symbols.end())
+    {
+      // Cast the address to the appropriate function type and call it,
+      // forwarding all arguments
+      return reinterpret_cast<Ret(*)(Args...)>(symbol->second)(args...);
+    }
+    else
+    {
+      throw std::runtime_error(std::string("Function not found: ") + name);
+    }
+  }
+
+  template <typename T>
+  T* GetVar(const char* name)
+  {
+    auto symbol = m_symbols.find(name);
+    if (symbol != m_symbols.end())
+    {
+      return reinterpret_cast<T*>(symbol->second);
+    }
+    else
+    {
+      // We didn't find the variable. Return an empty pointer
+      return nullptr;
```

I think throwing an exception here as well rather than returning nullptr would save user from having to make check nullptr checks everytime.